### ISSUE TITLE
Inconsistent Undo Behaviour

### ISSUE THREAD

**Issue by bbceg** [kewisch/quickmove-extension#16]
Version: 2.0.0pre1
Thunderbird: 68.3.0
OS: Windows 10

Steps to reproduce:

1. Move an email to a folder
2. Undo the action

Expected result:
Email reappears 

Actual result:
Performs the last "non-quickmove" action (e.g. restoring a deleted email)

**Comment by bbceg** [comment]
Not sure if this issue can be closed but perhaps worth noting that Undo works fine on v1.10.0 (as I found out to my delight just now!).

**Comment by kewisch** [comment]
Unfortunately this is broken again due to the rewrite. The new APIs don't have support for the undo/redo manager. See https://bugzilla.mozilla.org/show_bug.cgi?id=1671042 for details. I also don't see an easy way to do this with an internal API.

**Comment by bbceg** [comment]
Still seems to be working for me on v1.10 on Thunderbird 102.14.0, or does this only apply to v2?

**Comment by kewisch** [comment]
Yes, applies only to v2. I've started using the new WebExtensions/MailExtensions technology here, and those don't support the undo/redo stack at current.

**Comment by bbceg** [comment]
Okay, thanks for confirming and, as ever, your efforts with the add-on. Not sure when it makes to make the switch to v2, 1.10 is working well but presumably has a shelf life.

**Comment by kewisch** [comment]
Once you upgrade to Thunderbird 115, you'll necessarily be updated to v2. 115 is already out, so just a matter of when automatic updates are pushed out to everyone, or your distro doing the update.

**Comment by bbceg** [comment]
Rats, I was afraid you might say that! Hopefully the undo/redo manager will gain support in the not too distant future.

**Comment by ThomGr** [comment]
I understand the difficulty to implement a full undo mechanism, but this is very often causing an issue. If for whatever reason you accidentally move an email to the wrong folder (I have lots of those), and you do not remember a detail of the email, there is no realistic way to find it again (unless you know e.g. the date it was received, the sender etc.). So two suggestions:
1. can you implement a simple "undo the last action", so the extension itself remembers the last move (and only one, not the stack) and undoes is without attaching to the general undo mechanism. This would of course require a separate shortcut
2. alternatively, can you display the last move (subject and target folder) so there is a chance to find the last email moved, or implement a "go to" for that last email?

**Comment by liar666** [comment]
> I understand the difficulty to implement a full undo mechanism, but this is very often causing an issue. If for whatever reason you accidentally move an email to the wrong folder (I have lots of those), and you do not remember a detail of the email, there is no realistic way to find it again (unless you know e.g. the date it was received, the sender etc.). So two suggestions:
> 
>     1. can you implement a simple "undo the last action", so the extension itself remembers the last move (and only one, not the stack) and undoes is without attaching to the general undo mechanism. This would of course require a separate shortcut
> 
>     2. alternatively, can you display the last move (subject and target folder) so there is a chance to find the last email moved, or implement a "go to" for that last email?

+1

**Comment by jobisoft** [comment]
We added support for undo! Add-on developers can now declare message move/copy operations as user actions, which will behave as if the user did those operations in the UI.

https://bugzilla.mozilla.org/show_bug.cgi?id=1849476

`browser.messages.move(<ids>,<destFolder>, { isUserAction: true })`

**Comment by bbceg** [comment]
Excellent news, thank you @jobisoft and anyone else involved.

**Comment by kewisch** [comment]
Fixed in c681e5a566af20c5443dfd0d07967eabe732cd9e , thanks @jobisoft for taking care!
